cmd/crank/render/render_test.go

```diff
@@ -739,6 +739,207 @@ func TestRender(t *testing.T) {
+										AnnotationKeyRuntimeDevelopmentTarget: lis.Addr().String(),
+									},
+								},
+							}
+						}(),
+					},
+					Context: map[string][]byte{
+						"crossplane.io/context-key": []byte(`{}`),
+					},
+					ObservedResources: []composed.Unstructured{
+						{
+							Unstructured: unstructured.Unstructured{
+								Object: MustLoadJSON(`{
+									"apiVersion": "v1",
+									"kind": "Secret",
+									"metadata": {
+										"name": "secret1",
+										"namespace": "default"
+									},
+									"data": {
+										"foo": "Zm9v"
+									}
+								}`),
+							},
+						}, {
+							Unstructured: unstructured.Unstructured{
+								Object: MustLoadJSON(`{
+									"apiVersion": "v1",
+									"kind": "Secret",
+									"metadata": {
+										"name": "secret2",
+										"namespace": "default"
+									},
+									"data": {
+										"boo": "YnV6"
+									}
+								}`),
+							},
+						}, {
+							Unstructured: unstructured.Unstructured{
+								Object: MustLoadJSON(`{
+									"apiVersion": "btest.crossplane.io/v1",
+									"kind": "BComposed",
+									"metadata": {
+										"generateName": "test-render-",
+										"labels": {
+											"crossplane.io/composite": "test-render"
+										},
+										"annotations": {
+											"crossplane.io/composition-resource-name": "a-cool-resource"
+										},
+										"ownerReferences": [{
+											"apiVersion": "nop.example.org/v1alpha1",
+											"kind": "XNopResource",
+											"name": "test-render",
+											"blockOwnerDeletion": true,
+											"controller": true,
+											"uid": ""
+										}]
+									},
+									"spec": {
+										"widgets": 9002,
+										"writeConnectionSecretToRef": {
+											"name": "secret2",
+											"namespace": "default"
+										}
+									}
+								}`),
+							},
+						}, {
+							Unstructured: unstructured.Unstructured{
+								Object: MustLoadJSON(`{
+									"apiVersion": "btest.crossplane.io/v1",
+									"kind": "BComposed",
+									"metadata": {
+										"generateName": "test-render-",
+										"labels": {
+											"crossplane.io/composite": "test-render"
+										},
+										"annotations": {
+											"crossplane.io/composition-resource-name": "missing-connectionsecret"
+										},
+										"ownerReferences": [{
+											"apiVersion": "nop.example.org/v1alpha1",
+											"kind": "XNopResource",
+											"name": "test-render",
+											"blockOwnerDeletion": true,
+											"controller": true,
+											"uid": ""
+										}]
+									},
+									"spec": {
+										"widgets": 9002,
+										"writeConnectionSecretToRef": {
+											"name": "no-secret",
+											"namespace": "default"
+										}
+									}
+								}`),
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				out: Outputs{
+					// TODO: expect connectionDetails in composite when support is added
```

Comment:
@Avarei do you want to go ahead and tackle this part too? if i'm reading https://github.com/crossplane/crossplane/issues/4808 correctly, then it would be fully completed with what you've already done in this PR plus if you were to add connection details in the output. What do you think?

Reply:
Here you go :)
I've implemented it in https://github.com/crossplane/crossplane/pull/5985/commits/c87edf5aba954ddaeed904fb2303302947378234.

Reply:
I'm not sure how to add it as an e2e test, but I hope this helps: https://gist.github.com/Avarei/19bae623e8fe5290991698d4fb3cacd4